[DX-1852] Tyk Streams Improve Content (#6070)
TykTechnologies/tyk-docs: tyk-docs/content/api-management/event-driven-apis.md

Reviewer: file name can be `manage-event-driven-apis`

Author: Should I rename it? We need to add another alias for this change

  - /product-stack/tyk-streaming/guides/bloblang/methods/type-coercion/
---

<!-- ## TODO: Add availability -->

## Overview

*Tyk Streams* is a feature of Tyk API management platform that enables organizations to securely expose,
manage and monetize real-time event streams and asynchronous APIs.

With *Tyk Streams*, you can easily connect to event brokers and streaming platforms, such as
[Apache Kafka](https://github.com/TykTechnologies/tyk-pro-docker-demo/tree/kafka), and expose them as
managed API endpoints to internal and external consumers.

> <!-- TODO: Need to chanage Image -->

Reviewer 2: Assume that this needs fixing stil?

Author: Yes, i have added an image in the latest content.

full potential of their event-driven systems while maintaining the same level of security, control and visibility they
expect from their API management solution.

> ### Key Benefits

Reviewer: ```suggestion
### Why use Tyk Streams
```

> Tyk Streams seamlessly integrates with the Tyk API Gateway, allowing you to manage asynchronous APIs and event streams

Reviewer: Don't think we should use integrate. it's integral part of the gw

Author: This can be resolved, as this has been taken care of in your suggestion.

> Tyk Streams seamlessly integrates with the Tyk API Gateway, allowing you to manage asynchronous APIs and event streams

Reviewer: ```suggestion
Tyk Stream is a powerful stream processing engine integrated into the Tyk API Gateway, available as part of the Enterprise Edition. It allows you to manage asynchronous APIs and event streams as part of your API ecosystem, and enables real-time data processing, transformation, and intelligent routing.
```

Reviewer: something along these lines

- **Analytics**: Monitor the usage and performance of your async APIs with detailed analytics and reporting. Gain insights into consumer behavior and system health.
- **Developer Portal**: Publish your async APIs to the Tyk Developer Portal, providing a centralised catalog for discovery, documentation and subscription management.

### Supported Connectors and Protocols

Tyk Streams provides out-of-the-box connectors for popular event brokers and async protocols, including:

- [Apache Kafka](https://kafka.apache.org/documentation/)
- [WebSocket](https://websocket.org/guides/websocket-protocol/)
- [Server-Sent Events](https://en.wikipedia.org/wiki/Server-sent_events) (SSE)
- [Webhooks](https://en.wikipedia.org/wiki/Webhook)

## Quick Start

> 

Reviewer: We needs something like 
> This guide will help you implement your first event-driven API with Tyk in under 10 minutes.

Author: Added a similar line at the start of the quick start.

> To illustrate the capabilities of Tyk Streams, let's consider an example: building a basic asynchronous chat application, nicknamed **Chat Jippity**. In this scenario, a user sends a message (e.g., asking for a joke) via a simple web interface, and receives a response generated by a backend service asynchronously.

Author: @asoorm What do you think? I believe Tyk Streams is complex, and your (chatJippity) example demonstrates this nicely. Therefore, I propose renaming the header `Quick Start` to `Getting Started` to indicate it will be lengthier than a quick start.

> To illustrate the capabilities of Tyk Streams, let's consider an example: building a basic asynchronous chat application, nicknamed **Chat Jippity**. In this scenario, a user sends a message (e.g., asking for a joke) via a simple web interface, and receives a response generated by a backend service asynchronously.

Reviewer: We'll have mixed opinions whatever approach we'll take ;)

    *   **Decoupling:** The browser only needs to know the Tyk HTTP endpoint, not the Kafka details (brokers, topic name, protocol).
    *   **API Management:** Tyk can enforce authentication, rate limits, etc., on the `/chat` endpoint before the message even enters the Kafka system.

#### Use Case: Tyk as an API Consumer Gateway (Stream to Client)

*   **Goal:** Deliver the asynchronous response (the joke) from the backend system back to the client in real-time.
*   **Scenario:** The "Joker Service" has processed the request and generated a joke. It needs to send this back to the originating user's browser session.
*   **Flow:**
    1.  The Joker Service *publishes* the joke response as a message onto a different backend topic (e.g., the `discussion` topic in Kafka).
    2.  **Tyk Streams Role (Consumer):** Tyk Gateway is configured via another (or the same) API definition to *subscribe* to the `discussion` topic.
    3.  When Tyk receives a message from the `discussion` topic, it *pushes* the message content (the joke) to the appropriate client(s) using a suitable real-time protocol like Server-Sent Events (SSE) or WebSockets. 

>     **Note:** In case of multiple clients, events would round-robin amongst the consumers. So consumers would only receive every-other message.

Reviewer: Why? is it an example limitation?

Reviewer: The every other message is how tyk streams works?

- **Dashboard License**: We will configure Streams API using Dashboard. [Contact support](https://tyk.io/contact/) to obtain a license.
- **Curl and JQ**: These tools will be used for testing.

### Instructions

1. **Clone Git Repository:**

    The [tyk-demo](https://github.com/TykTechnologies/tyk-demo) repository offers a docker-compose environment you can run locally to explore Tyk streams. Open your terminal and clone the git repository using the below command.

    ```bash
    git clone https://github.com/TykTechnologies/tyk-demo
    cd tyk-demo
    ```

2. **Enable Tyk Streams:**

   By default, Tyk Streams is disabled. To enable Tyk Streams in the Gateway and Dashboard, you need to configure the following settings:

   Create an `.env` file and populate it with the values below:

   ```bash
   DASHBOARD_LICENCE=<your-dashboard-license>
   GATEWAY_IMAGE_REPO=tyk-gateway-ee
   TYK_DB_STREAMING_ENABLED=true
   TYK_GW_STREAMING_ENABLED=true
   ```

   - `DASHBOARD_LICENCE`: Add your license key. Contact [support](https://tyk.io/contact/) to obtain a license.

>    - `GATEWAY_IMAGE_REPO`: Tyk Streams is available in the Enterprise Edition of the Gateway.

Reviewer: ```suggestion
   - `GATEWAY_IMAGE_REPO`: Tyk Streams is available as part of the Enterprise Edition of the Gateway.
```

   - `TYK_DB_STREAMING_ENABLED` and `TYK_GW_STREAMING_ENABLED`: These must be set to `true` to enable Tyk Streams in the Dashboard and Gateway, respectively. Refer to the [configuration options]({{< ref "tyk-oss-gateway/configuration#streaming" >}}) for more details.

3. **Start Tyk Streams**

    Execute the following command:
    ```bash
    ./up.sh
    ```

    This process will take a couple of minutes to complete and will display some credentials upon completion. Copy the Dashboard **username, password, and API key**, and save them for later use.
    ```
            ▾ Tyk Demo Organisation
                  Username : admin-user@example.org
                  Password : 3LEsHO1jv1dt9Xgf
          Dashboard API Key : 5ff97f66188e48646557ba7c25d8c601
    ```

4. **Verify Setup:**

    Open Tyk Dashboard in your browser by visiting [http://localhost:3000](http://localhost:3000) or [http://tyk-dashboard.localhost:3000](http://tyk-dashboard.localhost:3000) and login with the provided **admin** credentials.

5. **Create Producer API:**

>     Create a file `producer.json` with the below content: (**Note:** Replace `<replace-with-your-ip>` with your IP assigned to your computer)

Reviewer: What does this mean `<replace-with-your-ip>`

        "paths": {},
        "servers": [
            {
                "url": "http://tyk-gateway.localhost:8080/gpt-chat/"
            }
        ],
        "x-tyk-api-gateway": {
            "info": {
                "name": "gpt-chat",
                "state": {
                    "active": true
                }
            },
            "server": {
                "listenPath": {
                    "value": "/gpt-chat/",
                    "strip": true
                }
            },
            "upstream": {
              "url": ""
            }
        },
        "x-tyk-streaming": {
            "streams": {
                "default_stream": {
                    "input": {
                        "http_server": {
                            "address": "",
                            "allowed_verbs": [
                                "POST"
                            ],
                            "path": "/chat",
                            "rate_limit": "",
                            "timeout": "5s"
                        },
                        "label": ""
                    },
                    "output": {
                        "kafka": {

>                             "addresses": ["<replace-with-your-ip>:9092"],

Reviewer: We need to explain that in the docker compose we started a kafka service 
```  kafka:
    image: apache/kafka:3.8.0
    networks:
      - tyk
    ports:
      - "9092:9092"```

Reviewer: This will cause users to fail. Per previous conversation- why not put the correct value here , correct for Mac or other machine and explain it. Or remove the the " " which make the copy/paste code work but then logically break the test

    ```

5. **Create Consumer API:**

    Create a file `consumer.json` with the below content: (**Note:** Replace `<replace-with-your-ip>` with your IP assigned to your computer)

    ```json
    {
        "components": {},
        "info": {
            "title": "gpt-discuss",
            "version": "1.0.0"
        },
        "openapi": "3.0.3",
        "paths": {},
        "servers": [
            {
                "url": "http://tyk-gateway.localhost:8080/gpt-discuss/"
            }
        ],
        "x-tyk-api-gateway": {
            "info": {
                "name": "gpt-discuss",
                "state": {
                    "active": true
                }
            },
            "server": {
                "listenPath": {
                    "value": "/gpt-discuss/",
                    "strip": true
                }
            },
            "upstream": {
              "url": ""
            }
        },
        "x-tyk-streaming": {
            "streams": {
                "default_stream": {
                    "input": {
                        "kafka": {

>                             "addresses": ["<replace-with-your-ip>:9092"],

Reviewer: This needs to break the curl if the user hasn't updated it. I just successfully run it and forgot to update the IP...

Tyk Streams is implemented as a middleware component in the Tyk Gateway. The architecture consists of the following logical components:

1. **Stream Middleware:** Think of this as the barista taking your order. It grabs incoming API requests and decides if they need streaming, slotting them into the right workflow after checking credentials and limits—just like making sure you’ve paid before handing over your latte.
2. **Stream Manager**: This is the shift supervisor. It oversees the creation and management of streaming tasks, starting or stopping them as needed, so everything runs like clockwork.
3. **Stream Instance**: These are the baristas at work. Each one handles a specific stream of data, processing it based on the instructions (configuration) it’s given.
4. **Stream Analytics**: Picture a chalkboard tracking how many coffees were made and how fast. This captures usage stats and performance metrics, feeding them into Tyk’s analytics system for a clear picture of what’s happening.

<div style="display: flex; justify-content: center;">

> {{< img src="/img/streams/tyk-streams-how-it-works.png" alt="How Tyk Streams work" width="500px" height="500px" >}}

Reviewer: How did you create the diagram? with Probe?

Author: Yes

> {{< img src="/img/streams/tyk-streams-how-it-works.png" alt="How Tyk Streams work" width="500px" height="500px" >}}

Reviewer: Is it possible to put the mermain code in the file, not the png?

Under the hood, Tyk Streams utilizes [Redis Streams](https://redis.io/docs/latest/develop/data-types/streams/) for efficient and reliable message distribution within the Tyk Gateway cluster. This enables:

- **Fault tolerance**: If a gateway node fails, clients connected to other nodes continue to receive messages uninterrupted.
- **Load balancing**: Messages are evenly distributed across the gateway cluster, ensuring optimal performance and resource utilization.
- **Message persistence**: Redis Streams provides durability, allowing messages to be persisted and replayed if necessary.

## Configuration Options

Tyk Streams is configured at both the system level (Tyk Gateway and Dashboard) and the API definition level. System-level settings control global streaming behavior, while API-level configurations define specific stream inputs and outputs for an API.

### System Level

#### Gateway Configuration

In the `tyk.conf` file, Streams is configured in the streaming section:

{{< tabs_start >}}

{{< tab_start "Config File" >}}

> ```json

Reviewer: ```suggestion
```yaml
```

...
"streaming": {
  "enabled": true,
}
...
```

{{< tab_end >}}

{{< tab_start "Environment Variable" >}}

```bash
TYK_GW_STREAMING_ENABLED=true
```

{{< tab_end >}}

{{< tabs_end >}}

This enables Tyk Streams on Gateway. Refer to the [Tyk Gateway Configuration Reference]({{< ref "tyk-oss-gateway/configuration#streamingenabled" >}}) for details.

#### Dashboard Configuration

In the `tyk_analytics.conf` file, Streams is configured in the streaming section:

{{< tabs_start >}}

{{< tab_start "Config File" >}}

```json
...
"streaming": {
  "enabled": true,
}
...
```

{{< tab_end >}}

{{< tab_start "Environment Variable" >}}

```bash
TYK_DB_STREAMING_ENABLED=true
```

{{< tab_end >}}

{{< tabs_end >}}

This enables Tyk Streams on Dashboard. Refer to the [Tyk Dashboard Configuration Reference]({{< ref "tyk-dashboard/configuration#streamingenabled" >}}) for details.

### API Definition

At the API definition level, Tyk Streams is configured through an OpenAPI extension `x-tyk-streaming`:

> ```json

Reviewer: ```suggestion
```yaml
```

"x-tyk-streaming": {
  "streams": {
    "stream1": {
      "input": { ... },
      "output": { ... }
    }
  }
}
```

{{< tabs_start >}}

{{< tab_start "Tyk OAS API (default)" >}}

> ```json

Reviewer: ```suggestion
```yaml
```

{{< tab_end >}}

{{< tab_start "Tyk Classic API" >}}

Tyk Streams is not supported for Tyk Classic API definition.

{{< tab_end >}}

{{< tab_start "Dashboard UI" >}}

1. Click on **Streams & Events** from the sidebar. This will open a form for creating the Streams API.  
   {{< img src="/img/streams/sidebar-navigation.png" alt="Sidebar Navigation" width="670px" height="500px" >}}
   {{< img src="/img/streams/streams-and-events-wizard.png" alt="Streams & Events Wizard" width="670px" height="500px" >}}

2. Enter a **unique API name**, select the **Streams** option, and click **Continue**.
   {{< img src="/img/streams/streams-option.png" alt="Streams Option" width="670px" height="500px" >}}

3. On the next screen, configure your Streams API:
   {{< img src="/img/streams/selection.png" alt="Output Selection" width="670px" height="500px" >}}

4. On the **API Details page**, review the auto-generated YAML configuration and **save** the API.
   {{< img src="/img/streams/api-details-page.png" alt="API Details Page" width="670px" height="500px" >}}

{{< tab_end >}}

{{< tabs_end >}}

> ## Use Cases

Reviewer: I think we need a line separator between H2s, as it's a different topic. What do you think

Author: Good idea, I have added it.

    *   **Decoupling:** The browser only needs to know the Tyk HTTP endpoint, not the Kafka details (brokers, topic name, protocol).
    *   **API Management:** Tyk can enforce authentication, rate limits, etc., on the `/chat` endpoint before the message even enters the Kafka system.

#### Tyk as an API Consumer Gateway (Stream to Client)

*   **Goal:** Deliver the asynchronous response (the joke) from the backend system back to the client in real-time.
*   **Scenario:** The "Joker Service" has processed the request and generated a joke. It needs to send this back to the originating user's browser session.
*   **Flow:**
    1.  The Joker Service *publishes* the joke response as a message onto a different backend topic (e.g., the `discussion` topic in Kafka).
    2.  **Tyk Streams Role (Consumer):** Tyk Gateway is configured via another (or the same) API definition to *subscribe* to the `discussion` topic.
    3.  When Tyk receives a message from the `discussion` topic, it *pushes* the message content (the joke) to the appropriate client(s) using a suitable real-time protocol like Server-Sent Events (SSE) or WebSockets. 

>     **Note:** In case of multiple clients, events would round-robin amongst the consumers. So consumers would only receive every-other message.

Author: ```suggestion
    **Note:** In case of multiple clients, events would round-robin amongst the consumers.```

```mermaid
sequenceDiagram
    participant Browser
    participant Tyk Gateway
    participant Joker Service

    Note over Browser, Tyk Gateway: Consumer - SSE Connection Setup
    Browser->>+Tyk Gateway: Make Server Side Events (SSE) Connection
    Tyk Gateway-->>-Browser: Connection Established

    Note over Browser, discKafka: Producer - Message Flow for Request/Response
    Browser->>+Tyk Gateway: (1) POST /chat (Request: "Tell me a joke")
    Tyk Gateway->>+chatKafka: (2) Publish message to 'chat' topic
    chatKafka-->>-Tyk Gateway: Ack (implied)
    chatKafka-->>+Joker Service: (3) Consume message from 'chat' topic
    Note over Joker Service: Processes request, gets joke
    Joker Service->>+discKafka: (4) Publish response to 'discussion' topic
    discKafka-->>-Joker Service: Ack (implied)
    discKafka-->>+Tyk Gateway: (6) Consume/Receive message from 'discussion' topic
    Tyk Gateway-->>-Browser: (7) Push joke response (via established WS/SSE connection)
    Tyk Gateway-->>-discKafka: Ack (implied)
```

Let's break down how Tyk Streams facilitates this, focusing on the distinct producer and consumer roles Tyk plays:

> ### Use Case

Reviewer: This title is confusing, especially since we have another one further below. We need to think of something clearer

Author: How about the following?

1. Example Scenario (I think this is good)
2. Application Overview
3. Practical Example
